reference/src/layout/unions.md

```diff
@@ -6,17 +6,98 @@ not to change until an RFC ratifies them.
 
 [#13]: https://github.com/rust-rfcs/unsafe-code-guidelines/issues/13
 
-The only degree of freedom the compiler has when computing the layout of a union
-like
+### Layout of individual union fields
 
-```rust,ignore
-union U { f1: T1, f2: T2 }
+The layout of each union field is determined by its type. 
+
+<details><summary><b>Rationale</b></summary>
+
+This is required to allow creating references to union fields:
+
+```rust
+# fn main() { unsafe {
+# #[derive(Copy, Clone)]
+struct T1;
+union U { f1: T1 }
+let u = U { f1: T1 };
+let t1: &T1 = &u.f1;
+// &T1 works for all references
+# }}
+```
+</details>
+
+### Unions with default layout ("`repr(Rust)`")
+
+**The default layout of unions is**, in general, **not specified.** 
+
+<details><summary><b>Rationale</b></summary>
+
+As of this writing, we want to keep the option of using non-zero offsets open
+for the future; whether this is useful depends on what exactly the
+compiler-assumed invariants about union contents are. This might become clearer
+after the validity of unions
+([unsafe-code-guidelines/73](https://github.com/rust-lang/unsafe-code-guidelines/issues/73))
+is settled.
+
+Even if the offsets happen to be all 0, there might still be differences in the
+function call ABI.  If you need to pass unions by-value across an FFI boundary,
+you have to use `#[repr(C)]`.
+
+</details>
+
+#### Layout of unions with a single non-zero-sized field
+
+The layout of unions with a single non-zero-sized field is the same as the
+layout of that field if:
+
+* that field has no padding bits, and
```

Comment:
We have not defined padding (what it is, how it behaves, where it occurs in which types) anywhere, and IMO it is not part of layout. So I feel like we are getting a little ahead of ourselves here.

Padding is IMO part of the value representation of a type; no interpretation of layout (size, align, niche, call ABI, field offsets) really includes "padding". For structs it naturally emerges as the "gaps between" fields, for enums I have no idea and for unions I am confused. ;)

Reply:
So we touched this on the meeting today. And it is not clear whether we have the definitions required to word this yet. We do not have a definition for padding, it is unclear whether a definition for padding is needed, but we could try to word this without using the "term" padding with the definitions that we have, and if we can't, we might be able to identify whatever definitions would be required to word something like this (and block this on working on those).